### DISCUSSION TITLE
Is there a variable which will tell me which log level is in effect during an agent run?

### DISCUSSION THREAD

**Discussion by craigcomstock** [cfengine/core#4801] [Q&A]
I want to pass along to a python script I am running, whether the agent is in inform, verbose or debug mode so that if a user is diagnosing a problem they can get more information.

I tried the following but always just get `R: args_list is --inform` instead of differences when I run with `--debug` `--verbose` and `--inform`.

```cf3
bundle agent main
{
  vars:
    "modes" slist => { "debug", "verbose", "inform" };
    "args_list" slist => { "--${modes}" },
      if => "${modes}_mode";

  reports:
    "args_list is ${args_list}";
}
```

**Comment by craigcomstock** [comment]
With some more experimentation and thinking I think I see the problem here. The "last one wins" when building up the slist with iterating over the values of `modes` as well as the fact that when in debug mode the agent also defines `verbose_mode` and `inform_mode` since those are also included. So if I change the order of the modes slist I get the answer I expect.

```cf3
bundle agent main
{
  vars:
    "modes" slist => { "inform", "verbose", "debug" };
    "args_list" slist => { "--${modes}" },
      if => "${modes}_mode";

  reports:
    "args_list is ${with}" with => ${args_list};
    inform_mode::
      "inform mode";
    debug_mode::
      "debug mode";
    verbose_mode::
      "verbose_mode";
}
```

```sh
$ cf-agent -Kdf ./test.cf | grep R:
R: args_list is --debug
R: inform mode
R: debug mode
R: verbose_mode

$ cf-agent -Kvf ./test.cf | grep R:
R: args_list is --verbose
R: inform mode
R: verbose_mode

$ cf-agent -KIf ./test.cf | grep R:
R: args_list is --inform
R: inform mode
```

**Comment by nickanderson** [comment]
Hi @craigcomstock,

I just wanted to offer another way to build a list based on class presence.

**Example policy:**

```
bundle agent __main__
{
  vars:
    # Here is our list of things
      "_modes" slist => { "inform", "verbose", "debug" };

    # Here we iterate over the list of things building up a classic associatve array if
      "_have_class[$(_modes)]"
        string => "$(_modes)_mode",
        if => "$(_modes)_mode";

    # Here we get the single list as requested
      "my_list_based_on_classes"
        slist => getvalues( _have_class );

  reports:
     # Print the list out after joining it together with commas for printing on a single line.
      "$(with)" with => join( ", ", my_list_based_on_classes );
}
```

**Example execution:**

```
chmod 600 /tmp/example.cf
cf-agent --no-lock --file /tmp/example.cf --log-level inform | grep "^R:"
cf-agent --no-lock --file /tmp/example.cf --log-level verbose | grep "^R:"
cf-agent --no-lock --file /tmp/example.cf --log-level debug | grep "^R:"
```

**Example execution output:**

```
R: inform_mode
R: verbose_mode, inform_mode
R: verbose_mode, debug_mode, inform_mode
```

**Reply by craigcomstock** [reply]
Sure. That makes sense. The root issue for me is that I was looking really for a variable and not a class which told me which log level was in effect and/or requested when cf-agent was started.

**Comment by craigcomstock** [comment]
I understand that the log level can vary according to attributes on a promise so what I am after here is any log level requested by the cf-agent arguments on the command line.

**Reply by nickanderson** [reply]
Yeah, I don't think there is any way to achieve this currently. At least I can't think of a _good_ way to do it. For this kind of thing it seems like a new `this` or `sys` variable would be desirable. Note, we have variables for the control bodies (but this isn't in there, and even if it were it wouldn't reflect the options from the cli).

A bad idea that came to mind would be to use `this.promiser_pid` and then see about extracting the options from `/proc/$(this.promiser_pid)/cmdline`

**Comment by atsaloli** [comment]
On Mon, Oct 3, 2022 at 10:17 AM Craig Comstock ***@***.***>
wrote:

> I understand that the log level can vary according to attributes on a
> promise so what I am after here is any log level requested by the cf-agent
> arguments on the command line.
>

What's the use case, Craig?
